### PR TITLE
[WIP] cuSTL Host/DeviceBuffers: Contructors (Pointers)

### PR DESCRIPTION
**Pull request by Flamefire**
This implements the cartBuffer functions correctly by calling constructors that actually take parameters. Allowing the default ctor is dangerous.

This also adds the missing move operators required for effective returning.

This also supersedes #1093 

- [x] Do not merge until rebased and changed according to PR #1096 (see: https://github.com/ComputationalRadiationPhysics/picongpu/pull/1094#issuecomment-139565612)
- [x] Test plugins using this

### CODE REVIEW

**Comment by Heikman** [comment]
Thanks a lot for cleaning the hacks! Looks fine to me.

**Comment by Flamefire** [comment]
Offline discussion with @psychocoderHPC : This has an issue with pitches: The pitch for 3D buffers is not always `pitchX * sizeY`. For example when the buffer is already a view of a bigger buffer. In that case one needs the actual pitches. So the constructors should take a size_t[dim-1] instead of a simple size_t. Currently this is not possible. @psychocoderHPC is working on a patch that adds these information to PMacc buffers which is also required for master. Therefore this patch should be rebased onto that other patch once it is merged and the ctors and conversion functions should be changed accordingly.

TLDR: Do not merge until rebased to pitch patch (no pun intended).

**Comment by bussmann** [comment]
gnihihihihi, 'pitch patch'

**Comment by Flamefire** [comment]
I'd vote +1 for naming the future PR "pitch patch"

**Comment by psychocoderHPC** [comment]
@Flamefire two comments a buff  you wrote `TLDR: Do not merge until rebased to pitch patch (no pun intended).` Could you please move it to the pull request description and add a issue number. It's hard to see why this pull request is blocked if the `do not merge` warning is inside the discussion flow.

**Comment by Flamefire** [comment]
Well its the 3rd comment, but ok. I added it to the description

**Comment by Flamefire** [comment]
Rebased and fixed for pitches. Tests pending but feel free to review this @Heikman @psychocoderHPC

**Comment by Flamefire** [comment]
Tested using the phasespace plugin for xpx, xpy, ypx, ypy, zpx, zpz at 1000 timesteps. Relative difference less than 0.04 and optically equivalent.

--> Ready to be merged
